### PR TITLE
Gh scraping

### PR DESCRIPTION
**Pull request by WojtekS-dev**
Changed the database structure - now users and repos are in separate tables with many-to-many connection between them
Added functions for scraping GH
Refactored filtering in API to work with new database

### CODE REVIEW

File: backend/env.txt [hunk 1]
```diff
@@ -1,2 +1,2 @@
 SECRET_KEY="SAMPLE_STRING"
-SQLALCHEMY_DATABASE_URI=mysql+pymysql://<db_user>:<db_password>@<db_host>/<db_name>?charset=utf8mb4
\ No newline at end of file
+SQLALCHEMY_DATABASE_URI=mysql+pymysql://user_dev:userdev@localhost/gh_db?charset=utf8mb4
```
Comment (PK109) [inline backend/env.txt]: This part shall not be changed

File: backend/app/GH_API_handling.py [hunk 2]
```diff
@@ -15,56 +17,251 @@
...
 
 
-def search_for_user(username=None):
+def user_is_in_db(usrname=None):
+    """Check if the user with usrname is in database"""
+    found_user = GithubUserInfo.query.filter_by(username=usrname).first()
 
-    g = Github(str(GH_access_token))
+    if found_user:
+        print(f"User {usrname} has been found in database")
```
Comment (orzeech) [inline backend/app/GH_API_handling.py]: use logging library instead of printing out

File: backend/app/GH_API_handling.py [hunk 3]
```diff
@@ -15,56 +17,251 @@
...
+            for contributor in repo.get_contributors():
+                if (
+                    contributor.login != "None"
+                    and user_is_in_db(contributor.login) == False
+                ):
+                    gh_user = add_user_to_database(
+                        contributor.login, return_added_user=True
+                    )
+                    # gh_user = GithubUserInfo(id=None, username=contributor.login)
+                    # db.session.add(gh_user)
```
Comment (orzeech) [inline backend/app/GH_API_handling.py]: do not commit commented-out code

File: backend/app/GH_API_handling.py [hunk 4]
```diff
@@ -15,56 +17,251 @@
...
+            repo_to_update.contributors = contributors_objs_list  # list of objects
+            try:
+                db.session.commit()
+                return f"Repo {repo_name} has been successfully updated"
 
-        if found_user:
-            return f"User {username} has been found in database"
+            except IntegrityError:
+                db.session.rollback()
+                return "something went wrong in update_repo"
```
Comment (orzeech) [inline backend/app/GH_API_handling.py]: log the error better

File: backend/app/GH_API_handling.py [hunk 5]
```diff
@@ -15,56 +17,251 @@
...
+        if repo_is_in_db(repo_name) and return_added_repo == False:
+            return f"Repo {repo_name} is already in database"
+        if repo_is_in_db(repo_name) and return_added_repo == True:
+            print(f"Repo {repo_name} is already in database")
+            return GithubRepositories.query.filter_by(reponame=repo_name).first()
+        else:
+            g = Github(str(GH_access_token))
+            repo = g.get_repo(repo_name)
+
+            # adds contibutors - turned off to prevent infinite loop
```
Comment (orzeech) [inline backend/app/GH_API_handling.py]: add a task to do this

File: backend/env.txt (hunk 1 above)
Comment (orzeech) [inline backend/env.txt]: these should probably be kept as is

**Review by orzeech — CHANGES_REQUESTED** [review]

**Review by orzeech — APPROVED** [review]

**Pull request merged**